### ISSUE TITLE
Dead tasks in ingress network after node failure

### ISSUE THREAD

**Issue by ZigmunD** [moby/moby#35548]
**Description**

Swarm cluster, 5 managers, 7 workers. One node stuck because of hw problems, we got dockerd zombie. I removed completelly that node, but ingress network still contains record for a task, that was on failed node and doesn't exist anymore. Fixed the node, wiped /var/lib/docker and added it to cluster.
```
zigmund@docker-m1:~$ docker service ps se-iagent-twemproxy 
ID                  NAME                    IMAGE                                                               NODE                DESIRED STATE       CURRENT STATE               ERROR               PORTS
u67qlmjjxnta        se-iagent-twemproxy.1   regisrty:5000/se/iagent-twemproxy:r1   docker-w5           Running             Running about an hour ago                       
gotnxegmpi5t        se-iagent-twemproxy.2   regisrty:5000/se/iagent-twemproxy:r1   docker-w4           Running             Running about an hour ago
zigmund@docker-m1:~$ docker network inspect -v ingress | jq ".[0].Services[\"se-iagent-twemproxy\"]"
{
  "VIP": "10.255.0.116",
  "Ports": [
    "Target: 6379, Publish: 10000"
  ],
  "LocalLBIndex": 297,
  "Tasks": [
    {
      "Name": "se-iagent-twemproxy.1.u67qlmjjxntapaewtk02ay83x",
      "EndpointID": "34ee3f7e1d331d00b01a8a86ce55d6d244dd680f1d9508e7e1a67cfc3ef15ac6",
      "EndpointIP": "10.255.0.118",
      "Info": {
        "Host IP": "10.21.241.55"
      }
    },
    {
      "Name": "se-iagent-twemproxy.2.gotnxegmpi5txme45mdu14kor",
      "EndpointID": "61391c797f8e0a3e572faffe2edf4456d108cfbe9d94987cc40049b5d7a32cf3",
      "EndpointIP": "10.255.0.117",
      "Info": {
        "Host IP": "10.21.241.54"
      }
    },
    {
      "Name": "se-iagent-twemproxy.2.172pra2m0c72s4bfyfmylqxhk",
      "EndpointID": "cdfd578825daac038dd2dfdd6c112ba5921e66c86c5d5766c5d0400b1b36ace4",
      "EndpointIP": "10.255.0.176",
      "Info": {
        "Host IP": "10.21.241.54"
      }
    }
  ]
}
```
As you can see, there are only 2 tasks exist, but ingress contains extra dead task (last one).

After service remove ingress still contains the service with only one dead task. Recreated the service -
 again extra task in ingress. Created service with different name - two services in ingress with same published port.

**Steps to reproduce the issue:**
1. Set up swarm
2. Brake some node??
3. Get stuck tasks in ingress.

**Describe the results you received:**
Ingress network containing dead tasks.

**Describe the results you expected:**
Ingress network without dead tasks.

**Additional information you deem important (e.g. issue happens only occasionally):**
Node add/rm, service del/create doesn't solves the problem.

**Output of `docker version`:**

```
Client:
 Version:      17.10.0-ce
 API version:  1.33
 Go version:   go1.8.3
 Git commit:   f4ffd25
 Built:        Tue Oct 17 19:04:16 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.10.0-ce
 API version:  1.33 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   f4ffd25
 Built:        Tue Oct 17 19:02:56 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 5
 Running: 5
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 17.10.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: gelf
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: uyish43royurno81qy40c9vcu
 Is Manager: false
 Node Address: 10.21.241.54
 Manager Addresses:
  10.21.241.56:2377
  10.21.241.57:2377
  10.21.241.58:2377
  10.21.241.59:2377
  10.21.241.60:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 0351df1c5a66838d0c392b4ac4cf9450de844e2d
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.11.0-14-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.67GiB
Name: docker-w4
ID: 3M77:QQVT:F32U:GFGX:N5ZN:6QCK:SYWP:RLEF:2X76:FDKW:VYNX:CBEI
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Ubuntu 16.04.3 LTS 4.11.0-14-generic, Proxmox VM.

**Comment by mborejdo** [comment]
We came across the same problem recently. 
After an unplanned reboot of one master node (currently running an all-manager setup in swarmmode due to problems with overlay-networks, as described here https://github.com/moby/moby/issues/32195) We currently have stale/wrong DNS entries for some services. 
I had to rename and restart the services to make them discoverable again.

Unfortunately I cannot find the stale overlay network entries on any node and have no way to remove these wrong IPs. Rebooting the node and rejoining the swarm does not help either.

We are currently running 17.10 but the problem appeared also earlier (we re-inited the swarm since then)

**Comment by mateuszpodraza** [comment]
We're experiencing the same issue here, but not just with the ingress network. Happens on other overlay networks as well.
We're doing a lot of joining, and draining then removing of nodes (scaling the swarm up and down, based on demand), so my best guess is that the two may be connected, because we can reproduce this on a daily basis.

Currently, we're running 17.09 (trying to upgrade to 17.10 was a disaster, and it seems 17.10 doesn't fix this) on CentOS 7.4 with the 4.14 kernel, 5 manager nodes in different availability zones and a bunch of workers (about 15 in idle, going up to a lot more when there's traffic).

There's about ~60 containers running in swarm when idle, getting up to hundreds when loaded. Happens on all types of containers that get created/destroyed frequently so I don't think there's a connection there.

What's weird is that some tasks in the network used to be on a node that both is no longer a part of the swarm nor even exists. We have not found a way to remove these tasks from the network, so we're currently experiencing major issues with connecting services.

By looking at the traffic using tcpdump, I've confirmed that all of our "host unreachable" and timeout errors are due to the routing mesh pointing to the already dead tasks - so some other issues could be connected to this one, I think.

Scaling the service down or up, and even removing and then recreating it does not get rid of the old entries.

I'd be happy to run some debug stuff if that helps you guys out, we're out of our depth here.

Pinging @ZigmunD @mborejdo @thaJeztah @abhi to hopefully stay in the loop

**Comment by akawula** [comment]
While connecting to the service I've got random timeouts. After few hours of debugging, I realized that this problem is laying in networking and dead tasks. Plz halp @ZigmunD  @mborejdo @abhi @fcrisciani !

**Comment by piotrmocko** [comment]
I am having the same problem as @mateuszpodraza. I have benig looking for a solution for a few weeks and I found this issue. My network is going creazy with those dead tasks. Do you know how to handle this problem?

**Comment by ddebroy** [comment]
I have been trying to repro this with 5 manager nodes and 7 worker nodes as mentioned in the original post and starting several replicas of nginx like this:

```
docker service create --name my-web2  --publish published=8081,target=80  --replicas 40 nginx
docker service create --name my-web3  --publish published=8082,target=80  --replicas 40 nginx
docker service create --name my-web4  --publish published=8083,target=80  --replicas 40 nginx
docker service create --name my-web5  --publish published=8084,target=80  --replicas 40 nginx
```

Next I tried to replicate an ungraceful shutdown on manager nodes using:
```
echo 1 > /proc/sys/kernel/sysrq
echo c > /proc/sysrq-trigger
```
as well as some node drains on the managers too.

After the above steps, I checked the number of tasks under `ingress` using:
```
docker network inspect -v ingress | jq ".[0].Services" | grep Name | grep my-web
```

So far I have not been able to reproduce this issue but will continue to try. For those who are seeing this issue, any chance you can provide some specific repro steps that may help narrow things down?

**Comment by ddebroy** [comment]
PR https://github.com/docker/libnetwork/pull/2013 addresses a race in this area that is a likely root cause for the issue. We can close this if the issue does not resurface in builds with the fix (>= CE 17.12).

**Comment by thaJeztah** [comment]
That PR was brought into master through https://github.com/moby/moby/pull/35620, and will be part of Docker 17.12.0; https://github.com/docker/docker-ce/blame/v17.12.0-ce-rc3/components/engine/vendor/github.com/docker/libnetwork/networkdb/cluster.go#L258


Let me close this issue for now, but comment if it's not resolved in 17.12 

/cc @fcrisciani
